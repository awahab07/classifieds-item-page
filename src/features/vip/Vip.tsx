import { Grid } from '@material-ui/core';
import React, { FunctionComponent, PropsWithChildren, ReactElement, useEffect } from 'react';
import { shallowEqual, useDispatch, useSelector } from 'react-redux';
import { IArticleImage } from '../../models/ArticleImage';
import { ImageResolution } from '../../models/ImageResolution';
import ImageSlider, { ISliderImageProps } from '../../shared/ImageSlider';
import { filterUniqueByUrl, getImageUrl } from '../../shared/utils/images';
import { useIsMobile } from '../../shared/utils/screen';
import { RootState } from '../../store';
import { vipActions } from './index';
import Slide from './Slide';
import { useStyles } from './styles';
import Thumbnail from './Thumbnail';

const Vip: FunctionComponent = (props: PropsWithChildren<{}>): ReactElement => {
  const dispatch = useDispatch();
  const { vip } = useSelector((state: RootState) => state.vipSlice, shallowEqual);
  useEffect(() => {
    dispatch(vipActions.getVip());
  }, []);
  const classes = useStyles();
  const isMobile = useIsMobile();

  const sliderImages: ISliderImageProps[] =
    vip?.images
      ?.map((image: IArticleImage) => ({
        thumbnailUrl: getImageUrl(image, ImageResolution.Thumbnail),
        url: getImageUrl(image, ImageResolution.Normal),
        fullScreenUrl: getImageUrl(image, ImageResolution.FullScreen),
        caption: vip.title,
      }))
      .filter(filterUniqueByUrl) ?? [];

  return (
    <Grid className={classes.grow} container direction={'column'} justify={'center'} alignItems={'center'}>
      <ImageSlider data={sliderImages} slideRenderer={Slide} thumbnailRenderer={Thumbnail} isMobile={isMobile} />
    </Grid>
  );
};

export default Vip;
