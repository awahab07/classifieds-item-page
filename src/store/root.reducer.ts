import { combineReducers } from '@reduxjs/toolkit';

import { vipReducer } from '../features/vip';

const rootReducer = combineReducers({
  vip: vipReducer,
});

export type RootState = ReturnType<typeof rootReducer>;

export default rootReducer;
