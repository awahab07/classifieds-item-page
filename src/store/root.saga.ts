import { all, fork } from 'redux-saga/effects';
import { vipListener } from '../features/vip';

export const rootSaga = function* root(): Generator {
  yield all([fork(vipListener)]);
};
