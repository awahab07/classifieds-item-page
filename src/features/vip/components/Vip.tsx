import { Grid, Paper } from '@material-ui/core';
import React, { FunctionComponent, PropsWithChildren, ReactElement, useEffect } from 'react';
import { shallowEqual, useDispatch, useSelector } from 'react-redux';
import { IArticleImage } from '../../../models/ArticleImage';
import { ImageResolution } from '../../../models/ImageResolution';
import ImageSlider, { ISliderImageProps } from '../../../shared/ImageSlider';
import { filterUniqueByUrl, getImageUrl } from '../../../shared/utils/images';
import { useIsMobile } from '../../../shared/utils/screen';
import { RootState } from '../../../store';
import { vipActions } from '../index';
import { useStyles } from '../styles';
import ArticleCta from './ArticleCta';
import MobileNavBar from './MobileNavBar';
import SellerInfo from './SellerInfo';
import Slide from './Slide';
import Thumbnail from './Thumbnail';
import VipHeader from './VipHeader';

const Vip: FunctionComponent = (props: PropsWithChildren<{}>): ReactElement => {
  const dispatch = useDispatch();
  const { vip } = useSelector((state: RootState) => state.vipSlice, shallowEqual);
  useEffect(() => {
    dispatch(vipActions.getVip());
  }, []);
  const classes = useStyles();
  const isMobile = useIsMobile();

  const sliderImages: ISliderImageProps[] =
    vip?.images
      ?.map((image: IArticleImage) => ({
        thumbnailUrl: getImageUrl(image, ImageResolution.Thumbnail),
        url: getImageUrl(image, ImageResolution.Normal),
        fullScreenUrl: getImageUrl(image, ImageResolution.FullScreen),
        caption: vip.title,
      }))
      .filter(filterUniqueByUrl) ?? [];

  return (
    <>
      {isMobile && vip !== null ? <MobileNavBar title={vip.title} price={vip.price} /> : null}
      <Paper elevation={0}>
        <Grid
          className={classes.grow}
          container
          direction={'column'}
          justify={'center'}
          alignItems={'center'}
        >
          {!isMobile && vip !== null ? <VipHeader {...vip} /> : null}
          <ImageSlider
            data={sliderImages}
            slideRenderer={Slide}
            thumbnailRenderer={Thumbnail}
            isMobile={isMobile}
          />
          {isMobile && vip !== null ? <ArticleCta {...vip} /> : null}
          {vip !== null ? <SellerInfo contact={vip.contact}/> : null}
        </Grid>
      </Paper>
    </>
  );
};

export default Vip;
