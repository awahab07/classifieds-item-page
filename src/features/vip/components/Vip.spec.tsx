import { mount } from 'enzyme';
import 'jest';
import React from 'react';
import { StaticRouter } from 'react-router';
import MOCK_VIP from '../../../../data/vip.json';
import { MockStoreProvider } from '../../../store/test';
import Vip from './Vip';

declare const global: any;

global.window = Object.create(window);
const url = 'http://test.com/category/product';
Object.defineProperty(window, 'location', {
  value: {
    href: url
  }
});

describe('Vip renders correctly', () => {
  it('should render correctly', () => {
    const Component = (
      <MockStoreProvider
        getState={() => ({ vipSlice: { vip: MOCK_VIP as any, loading: false, error: null } })}
      >
        <StaticRouter>
          <Vip />
        </StaticRouter>
      </MockStoreProvider>
    );
    const mounted = mount(Component);

    expect(mounted).toMatchSnapshot();
  });
});
