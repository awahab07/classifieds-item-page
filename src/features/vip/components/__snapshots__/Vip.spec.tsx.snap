// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Vip renders correctly should render correctly 1`] = `
<MockStoreProvider
  getState={[Function]}
>
  <Provider
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  >
    <StaticRouter>
      <Router
        history={
          Object {
            "action": "POP",
            "block": [Function],
            "createHref": [Function],
            "go": [Function],
            "goBack": [Function],
            "goForward": [Function],
            "listen": [Function],
            "location": Object {
              "hash": "",
              "pathname": "/",
              "search": "",
              "state": undefined,
            },
            "push": [Function],
            "replace": [Function],
          }
        }
        staticContext={Object {}}
      >
        <Vip>
          <Seo
            images={
              Array [
                Object {
                  "caption": "Ferrari 246 dino GT 1973",
                  "fullScreenUrl": "//i.ebayimg.com/00/s/NjgyWDEwMjQ=/z/znIAAOxynwlTcg-F/$_57.jpg",
                  "thumbnailUrl": "//i.ebayimg.com/00/s/NjgyWDEwMjQ=/z/znIAAOxynwlTcg-F/$_2.jpg",
                  "url": "//i.ebayimg.com/00/s/NjgyWDEwMjQ=/z/znIAAOxynwlTcg-F/$_27.jpg",
                },
                Object {
                  "caption": "Ferrari 246 dino GT 1973",
                  "fullScreenUrl": "//i.ebayimg.com/00/s/NjgyWDEwMjQ=/z/GvsAAOxyZzlTcg-L/$_57.jpg",
                  "thumbnailUrl": "//i.ebayimg.com/00/s/NjgyWDEwMjQ=/z/GvsAAOxyZzlTcg-L/$_2.jpg",
                  "url": "//i.ebayimg.com/00/s/NjgyWDEwMjQ=/z/GvsAAOxyZzlTcg-L/$_27.jpg",
                },
                Object {
                  "caption": "Ferrari 246 dino GT 1973",
                  "fullScreenUrl": "//i.ebayimg.com/00/s/NjgyWDEwMjQ=/z/GpoAAOxyZzlTcg-F/$_57.jpg",
                  "thumbnailUrl": "//i.ebayimg.com/00/s/NjgyWDEwMjQ=/z/GpoAAOxyZzlTcg-F/$_2.jpg",
                  "url": "//i.ebayimg.com/00/s/NjgyWDEwMjQ=/z/GpoAAOxyZzlTcg-F/$_27.jpg",
                },
                Object {
                  "caption": "Ferrari 246 dino GT 1973",
                  "fullScreenUrl": "//i.ebayimg.com/00/s/NjgyWDEwMjQ=/z/Z6UAAOxyCTtTcg-L/$_57.jpg",
                  "thumbnailUrl": "//i.ebayimg.com/00/s/NjgyWDEwMjQ=/z/Z6UAAOxyCTtTcg-L/$_2.jpg",
                  "url": "//i.ebayimg.com/00/s/NjgyWDEwMjQ=/z/Z6UAAOxyCTtTcg-L/$_27.jpg",
                },
              ]
            }
            url="http://test.com/category/product"
            vip={
              Object {
                "adMobTargeting": Object {
                  "advid": "2514",
                  "channel": "car",
                  "dealer": "01",
                  "ez": "2010",
                  "intid": "97045",
                  "kw": "193",
                  "ma": "179616",
                  "make": "Buick",
                  "model": "Skylark",
                  "price": "566",
                },
                "attributes": Array [
                  Object {
                    "label": "mileage",
                    "tag": "mileage",
                    "value": "179,616 km",
                  },
                  Object {
                    "label": "displacement",
                    "tag": "cubicCapacity",
                    "value": "1,320 cm³",
                  },
                  Object {
                    "label": "performance",
                    "tag": "power",
                    "value": "193 kW (262 hp)",
                  },
                  Object {
                    "label": "fuel type",
                    "tag": "fuel",
                    "value": "natural gas (CNG), E10-compatible, plug-in hybrid",
                  },
                  Object {
                    "label": "number of seats",
                    "tag": "numSeats",
                    "value": "2",
                  },
                  Object {
                    "label": "number of doors",
                    "tag": "doorCount",
                    "value": "6/7",
                  },
                  Object {
                    "label": "gear",
                    "tag": "transmission",
                    "value": "automatic",
                  },
                  Object {
                    "label": "pollutant class",
                    "tag": "emissionClass",
                    "value": "Euro2",
                  },
                  Object {
                    "label": "environmental badge",
                    "tag": "emissions sticker",
                    "value": "2 (red)",
                  },
                  Object {
                    "label": "first registration",
                    "tag": "firstRegistration",
                    "value": "08/2010",
                  },
                  Object {
                    "label": "air conditioning",
                    "tag": "climatisation",
                    "value": "automatic climate control",
                  },
                  Object {
                    "label": "airbags",
                    "tag": "airbag",
                    "value": "front airbags",
                  },
                  Object {
                    "label": "color",
                    "tag": "color",
                    "value": "black",
                  },
                  Object {
                    "label": "interior",
                    "value": "other, gray",
                  },
                ],
                "category": "sports car / coupe, day registration",
                "contact": Object {
                  "address1": "Rudower Chaussee 8",
                  "address2": "DE-46236 Bottrop Altstadt",
                  "canSendCcMail": true,
                  "country": "DE",
                  "enumType": "DEALER",
                  "imprint": "This is a default imprint text to ensure a correct test processing.",
                  "latLong": Object {
                    "lat": 51.48694,
                    "lon": 6.88274,
                  },
                  "name": "Handy Andy Home Improvement Center",
                  "phones": Array [
                    Object {
                      "number": "+4930123456789200",
                      "type": "PHONE1",
                      "uri": "tel: +4930123456789200",
                    },
                  ],
                  "type": "dealer",
                  "withMobileSince": "Dealer at mobile.de since 12/10/2014",
                },
                "created": 1418206195,
                "features": Array [
                  "towbar",
                  "handicapped accessible",
                  "hands-free system",
                  "navigation system",
                  "sports package",
                ],
                "hasDamage": false,
                "htmlDescription": "Ferrari Dino 246 Dino GT 20500 km. matching numbers Car has just done a full service at ferrari with 14.000 euro spent an it is in very good conditions. 20500 km. visible in London. Please no time wasters. if you wish to come and inspect it we have a pick up service for our valuable clients from london heathrow airport. we speak english italian russian polish lithuanian. car sold new in italy. ",
                "id": 97045,
                "images": Array [
                  Object {
                    "set": "fe4cfedffdffffffff",
                    "uri": "i.ebayimg.com/00/s/NjgyWDEwMjQ=/z/znIAAOxynwlTcg-F/$",
                  },
                  Object {
                    "set": "fe4cfedffdffffffff",
                    "uri": "i.ebayimg.com/00/s/NjgyWDEwMjQ=/z/znIAAOxynwlTcg-F/$",
                  },
                  Object {
                    "set": "fe4cfedffdffffffff",
                    "uri": "i.ebayimg.com/00/s/NjgyWDEwMjQ=/z/GvsAAOxyZzlTcg-L/$",
                  },
                  Object {
                    "set": "fe4cfedffdffffffff",
                    "uri": "i.ebayimg.com/00/s/NjgyWDEwMjQ=/z/znIAAOxynwlTcg-F/$",
                  },
                  Object {
                    "set": "fe4cfedffdffffffff",
                    "uri": "i.ebayimg.com/00/s/NjgyWDEwMjQ=/z/GpoAAOxyZzlTcg-F/$",
                  },
                  Object {
                    "set": "fe4cfedffdffffffff",
                    "uri": "i.ebayimg.com/00/s/NjgyWDEwMjQ=/z/GpoAAOxyZzlTcg-F/$",
                  },
                  Object {
                    "set": "fe4cfedffdffffffff",
                    "uri": "i.ebayimg.com/00/s/NjgyWDEwMjQ=/z/Z6UAAOxyCTtTcg-L/$",
                  },
                ],
                "isConditionNew": false,
                "isDamageCase": true,
                "isNew": false,
                "isPreRegistration": true,
                "makeId": 4400,
                "makeKey": "Ferrari",
                "modelKey": "246",
                "price": Object {
                  "grs": Object {
                    "amount": 566,
                    "currency": "EUR",
                  },
                  "large": "566 €",
                  "net": "476 €",
                  "nt": Object {
                    "amount": 475.63,
                    "currency": "EUR",
                  },
                  "type": "FIXED",
                  "vat": "7pm",
                },
                "segment": "Car",
                "sellerId": 56578,
                "title": "Ferrari 246 dino GT 1973",
                "vat": "19.00% VAT",
                "vc": "Car",
              }
            }
          >
            <HelmetWrapper
              defer={true}
              encodeSpecialCharacters={true}
            >
              <SideEffect(NullComponent)
                defer={true}
                encodeSpecialCharacters={true}
              >
                <NullComponent
                  defer={true}
                  encodeSpecialCharacters={true}
                />
              </SideEffect(NullComponent)>
            </HelmetWrapper>
            <HelmetWrapper
              defer={true}
              encodeSpecialCharacters={true}
            >
              <SideEffect(NullComponent)
                defer={true}
                encodeSpecialCharacters={true}
                meta={
                  Array [
                    Object {
                      "content": "Category: sports car / coupe, day registration, year: 246 details: Ferrari Dino 246 Dino GT 20500 km. matching numbers Car has just done a full service at ferrari with 14.000 euro spent an it is in very good conditions. 20500 km. visible in London. Please no time wasters. if you wish to come and inspect it we have a pick up service for our valuable clients from london heathrow airport. we speak english italian russian polish lithuanian. car sold new in italy. }",
                      "name": "description",
                    },
                    Object {
                      "content": "246,Ferrari,towbar,handicapped accessible,hands-free system,navigation system,sports package",
                      "name": "keywords",
                    },
                  ]
                }
                script={
                  Array [
                    Object {
                      "innerHTML": "{\\"@context\\":\\"https://schema.org/\\",\\"@type\\":\\"Product\\",\\"name\\":\\"Ferrari 246 dino GT 1973\\",\\"image\\":[\\"https://i.ebayimg.com/00/s/NjgyWDEwMjQ=/z/znIAAOxynwlTcg-F/$_57.jpg\\",\\"https://i.ebayimg.com/00/s/NjgyWDEwMjQ=/z/GvsAAOxyZzlTcg-L/$_57.jpg\\",\\"https://i.ebayimg.com/00/s/NjgyWDEwMjQ=/z/GpoAAOxyZzlTcg-F/$_57.jpg\\",\\"https://i.ebayimg.com/00/s/NjgyWDEwMjQ=/z/Z6UAAOxyCTtTcg-L/$_57.jpg\\"],\\"description\\":\\"Ferrari Dino 246 Dino GT 20500 km. matching numbers Car has just done a full service at ferrari with 14.000 euro spent an it is in very good conditions. 20500 km. visible in London. Please no time wasters. if you wish to come and inspect it we have a pick up service for our valuable clients from london heathrow airport. we speak english italian russian polish lithuanian. car sold new in italy. \\",\\"sku\\":\\"97045\\",\\"mpn\\":\\"4400\\",\\"brand\\":{\\"@type\\":\\"Brand\\",\\"name\\":\\"Ferrari\\"},\\"offers\\":{\\"@type\\":\\"Offer\\",\\"url\\":\\"http://test.com/category/product\\",\\"priceCurrency\\":\\"EUR\\",\\"price\\":\\"475.63\\",\\"itemCondition\\":\\"https://schema.org/UsedCondition\\",\\"availability\\":\\"https://schema.org/InStock\\",\\"seller\\":{\\"@type\\":\\"dealer\\",\\"name\\":\\"Handy Andy Home Improvement Center\\"}}}",
                      "type": "application/ld+json",
                    },
                  ]
                }
                title="Mobile.de | Cars | Ferrari 246 dino GT 1973"
                titleAttributes={Object {}}
              >
                <NullComponent
                  defer={true}
                  encodeSpecialCharacters={true}
                  meta={
                    Array [
                      Object {
                        "content": "Category: sports car / coupe, day registration, year: 246 details: Ferrari Dino 246 Dino GT 20500 km. matching numbers Car has just done a full service at ferrari with 14.000 euro spent an it is in very good conditions. 20500 km. visible in London. Please no time wasters. if you wish to come and inspect it we have a pick up service for our valuable clients from london heathrow airport. we speak english italian russian polish lithuanian. car sold new in italy. }",
                        "name": "description",
                      },
                      Object {
                        "content": "246,Ferrari,towbar,handicapped accessible,hands-free system,navigation system,sports package",
                        "name": "keywords",
                      },
                    ]
                  }
                  script={
                    Array [
                      Object {
                        "innerHTML": "{\\"@context\\":\\"https://schema.org/\\",\\"@type\\":\\"Product\\",\\"name\\":\\"Ferrari 246 dino GT 1973\\",\\"image\\":[\\"https://i.ebayimg.com/00/s/NjgyWDEwMjQ=/z/znIAAOxynwlTcg-F/$_57.jpg\\",\\"https://i.ebayimg.com/00/s/NjgyWDEwMjQ=/z/GvsAAOxyZzlTcg-L/$_57.jpg\\",\\"https://i.ebayimg.com/00/s/NjgyWDEwMjQ=/z/GpoAAOxyZzlTcg-F/$_57.jpg\\",\\"https://i.ebayimg.com/00/s/NjgyWDEwMjQ=/z/Z6UAAOxyCTtTcg-L/$_57.jpg\\"],\\"description\\":\\"Ferrari Dino 246 Dino GT 20500 km. matching numbers Car has just done a full service at ferrari with 14.000 euro spent an it is in very good conditions. 20500 km. visible in London. Please no time wasters. if you wish to come and inspect it we have a pick up service for our valuable clients from london heathrow airport. we speak english italian russian polish lithuanian. car sold new in italy. \\",\\"sku\\":\\"97045\\",\\"mpn\\":\\"4400\\",\\"brand\\":{\\"@type\\":\\"Brand\\",\\"name\\":\\"Ferrari\\"},\\"offers\\":{\\"@type\\":\\"Offer\\",\\"url\\":\\"http://test.com/category/product\\",\\"priceCurrency\\":\\"EUR\\",\\"price\\":\\"475.63\\",\\"itemCondition\\":\\"https://schema.org/UsedCondition\\",\\"availability\\":\\"https://schema.org/InStock\\",\\"seller\\":{\\"@type\\":\\"dealer\\",\\"name\\":\\"Handy Andy Home Improvement Center\\"}}}",
                        "type": "application/ld+json",
                      },
                    ]
                  }
                  title="Mobile.de | Cars | Ferrari 246 dino GT 1973"
                  titleAttributes={Object {}}
                />
              </SideEffect(NullComponent)>
            </HelmetWrapper>
          </Seo>
          <Switch>
            <Redirect
              computedMatch={
                Object {
                  "isExact": true,
                  "params": Object {},
                  "path": "/",
                  "url": "/",
                }
              }
              exact={true}
              from="/"
              location={
                Object {
                  "hash": "",
                  "pathname": "/",
                  "search": "",
                  "state": undefined,
                }
              }
              to="//Ferrari-246-dino-GT-1973"
            />
          </Switch>
        </Vip>
      </Router>
    </StaticRouter>
  </Provider>
</MockStoreProvider>
`;
